source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -172,56 +219,47 @@ const std::string& HeaderToMetadataFilter::decideNamespace(const std::string& ns
   return nspace.empty() ? HttpFilterNames::get().HeaderToMetadata : nspace;
 }
 
+// add metadata['key']= value depending on header present or missing case
+void HeaderToMetadataFilter::applyKeyValue(std::string value,
+                                           const Rule& rule,
+                                           const KeyValuePair& keyval,
+                                           StructMap& np) {
+  if (!keyval.value().empty()) {
+    value = keyval.value();
+  } else {
+    const auto& matcher = rule.regexRewrite();
+    if (matcher != nullptr) {
+      value = matcher->replaceAll(value, rule.regexSubstitution());
+    }
+  }
+  if (!value.empty()) {
+    const auto& nspace = decideNamespace(keyval.metadata_namespace());
+    addMetadata(np, nspace, keyval.key(), value, keyval.type(), keyval.encode());
+  } else {
+    ENVOY_LOG(debug, "value is empty, not adding metadata");
+  }
+}
+
 void HeaderToMetadataFilter::writeHeaderToMetadata(Http::HeaderMap& headers,
                                                    const HeaderToMetadataRules& rules,
                                                    Http::StreamFilterCallbacks& callbacks) {
   StructMap structs_by_namespace;
 
   for (const auto& rule : rules) {
-    const auto& header = rule.header();
     const auto& proto_rule = rule.rule();
-    const Http::HeaderEntry* header_entry = headers.get(header);
-
-    if (header_entry != nullptr && proto_rule.has_on_header_present()) {
-      const auto& keyval = proto_rule.on_header_present();
-      absl::string_view value = header_entry->value().getStringView();
-      // This is used to hold the rewritten header value, so that it can
-      // be bound to value without going out of scope.
-      std::string rewritten_value;
-
-      if (!keyval.value().empty()) {
-        value = absl::string_view(keyval.value());
-      } else {
-        const auto& matcher = rule.regexRewrite();
-        if (matcher != nullptr) {
-          rewritten_value = matcher->replaceAll(value, rule.regexSubstitution());
-          value = rewritten_value;
-        }
-      }
-
-      if (!value.empty()) {
-        const auto& nspace = decideNamespace(keyval.metadata_namespace());
-        addMetadata(structs_by_namespace, nspace, keyval.key(), value, keyval.type(),
-                    keyval.encode());
-      } else {
-        ENVOY_LOG(debug, "value is empty, not adding metadata");
-      }
-
-      if (proto_rule.remove()) {
-        headers.remove(header);
-      }
+    bool has_value;
```

Comment:
nit: just `value` and `has_value` inline and map `has_value` const. However, you can also just use `value.has_value()` or even just `value` (which will be true if set).